### ISSUE TITLE
vips in the browser

### ISSUE THREAD

**Issue by andrzejbl** [libvips/libvips#192]
Hi, 

I'm searching for a lib that will allow me to use minimum amount of memory while downsizing images.
I want to compile it to javascript so it will run in browser. It means that disk I/O also uses memory. 

Could you tell be what is the lowest memory usage when I'm resizing 36mpix image?

**Comment by cancerberoSgx** [comment]
Did anybody has success building a webassembly version of the library ? I'm the author of a ImageMagick webasesembly port using emscripten and there I compiled almost all of its delegate libraries some of which are also dependencies of this library - before I give it a try I would like to know if you think there is viable or if anybody has already tried or any experience, thanks!

**Comment by jcupitt** [comment]
Hi @cancerberoSgx, 

It has been talked about a few times, but I don't know of anyone who has actually done it. Please post an update if you get it working, I'd be interested.

**Comment by kleisauke** [comment]
I started working on this as part of a university research project. 

After 3 days of trial and error, I managed to create a working proof of concept:
https://kleisauke.nl/wasm-vips/

It's able to thumbnail a [4120x2747 image](https://kleisauke.nl/wasm-vips/zebra.jpg) to 500x333 within ~200 ms, which is quite nice. I compiled libvips using Emscripten with a minimal set of dependencies (gettext, zlib, ffi, glib, expat and webp). 

Note that this example uses the SharedArrayBuffer capability in browsers. That API allows sharing memory between the main thread and web workers as well as atomic operations for synchronization, which enables Emscripten to implement support for the Pthreads (POSIX threads) API.
It may not work in all browsers, and due to the recent Spectre and Meltdown vulnerabilities, browsers other than Chrome [have disabled SharedArrayBuffer by default](https://caniuse.com/#feat=sharedarraybuffer).

Most of the work consisted of patching the dependencies. I was able to compile libffi with [this patch](https://github.com/atrosinenko/libffi/compare/master...atrosinenko:emscripten), glib requires some minor patches in the build system. libvips itself requires a major patch (> 1100 lines) to [fix all function pointer issues](https://emscripten.org/docs/porting/guidelines/function_pointer_issues.html), for example:

<details><summary>Details</summary>

````patch
diff --git a/libvips/iofuncs/gate.c b/libvips/iofuncs/gate.c
index 1111111..2222222 100644
--- a/libvips/iofuncs/gate.c
+++ b/libvips/iofuncs/gate.c
@@ -210,8 +210,8 @@ vips__thread_profile_init_cb( VipsThreadProfile *profile )
 	vips_thread_profile_free( profile );
 }
 
-static void
-vips__thread_profile_init( void )
+static void *
+vips__thread_profile_init( void *dummy )
 {
 #ifdef HAVE_PRIVATE_INIT
 	static GPrivate private = 
@@ -223,6 +223,7 @@ vips__thread_profile_init( void )
 		vips_thread_profile_key = g_private_new( 
 			(GDestroyNotify) vips__thread_profile_init_cb );
 #endif
+	return( NULL );
 }
 
 static VipsThreadGate *
@@ -245,7 +246,7 @@ vips__thread_profile_attach( const char *thread_name )
 
 	VipsThreadProfile *profile;
 
-	VIPS_ONCE( &once, (GThreadFunc) vips__thread_profile_init, NULL );
+	VIPS_ONCE( &once, vips__thread_profile_init, NULL );
 
 	VIPS_DEBUG_MSG( "vips__thread_profile_attach: %s\n", thread_name ); 
 ````
</details>

It is undefined behavior in C and C++ to cast a function pointer to another type and call it that way. This [does work in most native platforms](https://stackoverflow.com/a/14044244), however, despite it being UB, but in WASM it fails on runtime.

The source code, build script and the necessary patches will be published soon. I intend to expand this proof of concept further into a "full-fledged vips playground", where all the features of libvips can be viewed and executed in the browser (I have about 10 weeks to spend on this project).

**Comment by jcupitt** [comment]
Wow! Very nice.

Yes, I think I used to put the cast in the call, eg.:

```C
 		vips_thread_profile_key = g_private_new( 
 			(GDestroyNotify) vips__thread_profile_init_cb );
```

Because it means you can write a more meaningful (to humans) signature for the function prototype. 

More recent code mostly gets rid of the cast in the call and adds a set of casts to the function instead, which is what I guess your patch does, Kleis. It looks uglier (to humans) but is better for machine type checkers, and obviously better for transpilers.

Anyway, this sounds great!

**Comment by ascorbic** [comment]
Very cool. This wouldn't just be good for using in the browser: this could allow a WASI port of Sharp, saving a lot of node-gyp pain. https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/

**Comment by wis** [comment]
here is a list of libvips' dependencies:
- cfitsio [A library of C and Fortran subroutines for reading and writing data files in FITS (Flexible Image Transport System) data format]
- fftw [A library for computing the discrete Fourier transform (DFT)]
- imagemagick [An image viewing/manipulation program]
- libexif [A library to parse an EXIF file and read the data from those tags]
- libgsf [An extensible I/O abstraction library for dealing with structured file formats]
- libheif [HEIF file format decoder and encoder]
- libimagequant [Library for high-quality conversion of RGBA images to 8-bit indexed-color (palette) images]
- librsvg [SVG rendering library]
- libwebp [WebP library and conversion tools]
- libxml2 [XML parsing library, version 2]
- openexr [An high dynamic-range image file format library]
- orc [Optimized Inner Loop Runtime Compiler]
- pango [A library for layout and rendering of text]
- poppler-glib [Poppler glib bindings]

does sharp not use any of the functionalities that depend on any of these dependencies and thus it can be/was removed?
which dependencies do you think is gonna be the hardest to port/compile with emscripten?
it looks like [_limited_ SIMD support is coming to WASM](https://v8.dev/features/simd) do you think that only Fixed-Width 128-bit SIMD operations only and no longer/wider width operations can improve on this result ?

**Comment by kleisauke** [comment]
Not all the dependencies you've listed are necessary. Only zlib, libffi, glib and libexpat are needed for a minimal build. 

For support for different image formats you can add libjpeg-turbo ([compiled without SIMD support](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/250#issuecomment-407615180), unfortunately), libwebp, libpng, giflib, and libtiff. For EXIF metadata in JPEG files and ICC profile support add libexif and lcms2.

Adding librsvg will be harder because it adds these (sub-)dependencies (+ Rust code is known to be large):
<details><summary>Details</summary>

- gdk-pixbuf
- cairo
  - pixman
  - fontconfig (recommend)
- pango
  - cairo (recommend)
  - fontconfig
  - freetype
  - harfbuzz
  - fribidi
- libxml2
</details>

liborc would be nice to have, but it does not support SIMD code within WASM (AFAIK).

These dependencies are optional and probably not needed within WASM environments:
<details><summary>Details</summary>

- libgsf (support for creating image pyramids with `dzsave`)
- fftw3 (support for fourier transforms) - _GPL license_
- imagemagick (support for loading all libMagick-supported image file types) - _GPL-like license_
- libheif (support for loading and saving HEIC images) - _patent-encumbered_
  - x265 (support for encoding to HEIC)  - _GPL license, patent-encumbered_
- libimagequant (support for writing 8-bit palette-ised PNGs) - _GPL license_
- poppler (support for loading PDF files) - _GPL license_
- pdfium (support for loading PDF files)
- pangoft2 (support for text rendering)
- matio (support for loading Matlab files)
- cfitsio (support for loading FITS images)
- libniftiio (support for loading and saving NIfTI images)
- openexr (support for loading OpenEXR images)
- openslide (support for loading OpenSlide images)
</details>

(I won't compile any dependencies noted with _GPL license_, _GPL-like license_ or _patent-encumbered_ due to license/patent complexity).

Answering your questions:
1. sharp doesn't use the optional dependencies listed above (except for libgsf, see [here](https://github.com/lovell/sharp-libvips/blob/d0a573b8d033fba79063d77e7b209b6c4d57130b/build/lin.sh#L19-L43)). But it does use liborc and librsvg (+ the sub-dependencies).
2. liborc and librsvg would be the hardest, I think.
3. It can improve the result. The (old) demo uses WebP images that are somehow slow within WASM, with a compiled libjpeg-turbo it is about ~12 ms(!).

**Comment by kleisauke** [comment]
This is now done! :tada:
https://libvips.github.io/libvips/2020/09/01/libvips-for-webassembly.html
https://github.com/kleisauke/wasm-vips

Here is a simple Node.js application to get you started:
```js
const Vips = require('wasm-vips');

(async () => {
    // Initialize wasm-vips
    const vips = await Vips();

    // Smart crop to 500x500 using the attention strategy
    // (searches the image for features which might catch a human eye)
    // Image source: https://www.flickr.com/photos/jasonidzerda/3987784466
    const thumbnail = vips.Image.thumbnail('owl.jpg', 500, {
        height: 500,
        crop: vips.Interesting.attention /*'attention'*/
    });

    // Write the result to a WebP file
    thumbnail.writeToFile('owl-thumb.webp');

    // Memory management
    thumbnail.delete();

    // Exit the Emscripten runtime (i.e. a more "commandline" experience)
    vips._exit(0);
})();
```

```bash
node --experimental-wasm-threads --experimental-wasm-simd --experimental-wasm-bulk-memory demo.js
```

Check out the web-based playground for more demos:
https://kleisauke.github.io/wasm-vips/playground

Issue https://github.com/kleisauke/wasm-vips/issues/1 and the matching [Kanban board](https://github.com/kleisauke/wasm-vips/projects/1) keeps track of what needs to be done before we can declare a stable release.

Feel free to report any issues to the wasm-vips issue tracker.
https://github.com/kleisauke/wasm-vips/issues

**Comment by amio** [comment]
Awesome! Seems we can continue with https://github.com/lovell/sharp/issues/1476 now @lovell
